crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cmp::Ordering, sync::Arc, time::Duration};
+use sui_config::node::AuthorityStorePruningConfig;
+use sui_types::base_types::{ObjectID, SequenceNumber, VersionNumber};
+use tokio::{
+    sync::oneshot::{self, Sender},
+    time::{self, Instant},
+};
+use tracing::log::{error, info};
+use typed_store::Map;
+
+use crate::authority::authority_store::ObjectKey;
+
+use super::authority_store_tables::AuthorityPerpetualTables;
+
+const MAX_OPS_IN_ONE_WRITE_BATCH: u64 = 10000;
+
+pub struct AuthorityStorePruner {
+    _objects_pruner_cancel_handle: oneshot::Sender<()>,
+}
+
+impl AuthorityStorePruner {
+    fn prune_objects(
+        num_versions_to_retain: u64,
+        perpetual_db: Arc<AuthorityPerpetualTables>,
+    ) -> u64 {
+        let iter = perpetual_db.objects.iter().skip_to_last().reverse();
```

Comment:
How does this iteration work internally? Does it make a copy of the entire table? If not, how does it handle concurrent writes/deletions?

Reply:
RocksDB iterator work on an immutable snapshot of the database which a view of the db at point in time. Snapshots are tied to the WAL sequence number and are cheap to create as db has to only filter out results with higher sequence number. Concurrent write/deletions are not visible to the iterator once it is created.

Reply:
This really is last writer wins scenario. If the pruner is the last one, it will end up deleting `(Object, Version0, NewContent) `or else it deletes `(Object, Version0, Content)` and we end up `(Object, Version0, NewContent)`. In the current scenario, we only append to the objects table with newer versions for an object id (and never really try to mutate older versions of an object id), so we should not run into this kind of situation i think.